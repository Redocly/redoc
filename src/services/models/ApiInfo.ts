import { OpenAPIContact, OpenAPIInfo, OpenAPILicense } from '../../types';
import { IS_BROWSER } from '../../utils/';
import { OpenAPIParser } from '../OpenAPIParser';

export class ApiInfoModel implements OpenAPIInfo {
  title: string;
  version: string;

  description: string;
  summary: string;
  termsOfService?: string;
  contact?: OpenAPIContact;
  license?: OpenAPILicense;

  downloadLink?: string;
  downloadFileName?: string;

  constructor(private parser: OpenAPIParser) {
    Object.assign(this, parser.spec.info);
    this.description = parser.spec.info.description || '';
    this.summary = parser.spec.info.summary || '';

    const firstHeadingLinePos = this.description.search(/^##?\s+/m);
    if (firstHeadingLinePos > -1) {
      this.description = this.description.substring(0, firstHeadingLinePos);
    }

    this.downloadLink = this.getDownloadLink();
    this.downloadFileName = this.getDownloadFileName();
  }

  private getDownloadLink(): string | undefined {
    if (this.parser.specUrl) {
      return this.parser.specUrl;
    }

    if (IS_BROWSER && window.Blob && window.URL && window.URL.createObjectURL) {
      const blob = new Blob([JSON.stringify(this.parser.spec, null, 2)], {
        type: 'application/json',
      });
      return window.URL.createObjectURL(blob);
    }
  }

  private getDownloadFileName(): string | undefined {
    if (!this.parser.specUrl) {
      return 'swagger.json';
    }
    return undefined;
  }
}
