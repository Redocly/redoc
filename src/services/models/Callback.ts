import { action, observable, makeObservable } from 'mobx';

import { OpenAPICallback, Referenced } from '../../types';
import { isOperationName, JsonPointer } from '../../utils';
import { OpenAPIParser } from '../OpenAPIParser';
import { OperationModel } from './Operation';
import { RedocNormalizedOptions } from '../RedocNormalizedOptions';

export class CallbackModel {
  @observable
  expanded: boolean = false;

  name: string;
  operations: OperationModel[] = [];

  constructor(
    parser: OpenAPIParser,
    name: string,
    infoOrRef: Referenced<OpenAPICallback>,
    pointer: string,
    options: RedocNormalizedOptions,
  ) {
    makeObservable(this);

    this.name = name;
    const paths = parser.deref<OpenAPICallback>(infoOrRef);
    parser.exitRef(infoOrRef);

    for (const pathName of Object.keys(paths)) {
      const path = paths[pathName];
      const operations = Object.keys(path).filter(isOperationName);
      for (const operationName of operations) {
        const operationInfo = path[operationName];

        const operation = new OperationModel(
          parser,
          {
            ...operationInfo,
            pathName,
            pointer: JsonPointer.compile([pointer, name, pathName, operationName]),
            httpVerb: operationName,
            pathParameters: path.parameters || [],
            pathServers: path.servers,
          },
          undefined,
          options,
          true,
        );

        this.operations.push(operation);
      }
    }
  }

  @action
  toggle() {
    this.expanded = !this.expanded;
  }
}
