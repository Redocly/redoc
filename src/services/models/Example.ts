import type { OpenAPIEncoding, OpenAPIExample, Referenced } from '../../types';
import { isFormUrlEncoded, isJsonLike, urlFormEncodePayload } from '../../utils/openapi';
import type { OpenAPIParser } from '../OpenAPIParser';

const externalExamplesCache: { [url: string]: Promise<any> } = {};

export class ExampleModel {
  value: any;
  summary?: string;
  description?: string;
  externalValueUrl?: string;

  constructor(
    parser: OpenAPIParser,
    infoOrRef: Referenced<OpenAPIExample>,
    public mime: string,
    encoding?: { [field: string]: OpenAPIEncoding },
  ) {
    const { resolved: example } = parser.deref(infoOrRef);
    this.value = example.value;
    this.summary = example.summary;
    this.description = example.description;
    if (example.externalValue) {
      this.externalValueUrl = new URL(example.externalValue, parser.specUrl).href;
    }

    if (isFormUrlEncoded(mime) && this.value && typeof this.value === 'object') {
      this.value = urlFormEncodePayload(this.value, encoding);
    }
  }

  getExternalValue(mimeType: string): Promise<any> {
    if (!this.externalValueUrl) {
      return Promise.resolve(undefined);
    }

    if (this.externalValueUrl in externalExamplesCache) {
      return externalExamplesCache[this.externalValueUrl];
    }

    externalExamplesCache[this.externalValueUrl] = fetch(this.externalValueUrl).then(res => {
      return res.text().then(txt => {
        if (!res.ok) {
          return Promise.reject(new Error(txt));
        }

        if (isJsonLike(mimeType)) {
          try {
            return JSON.parse(txt);
          } catch (e) {
            return txt;
          }
        } else {
          return txt;
        }
      });
    });

    return externalExamplesCache[this.externalValueUrl];
  }
}
