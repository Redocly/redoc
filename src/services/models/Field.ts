import { action, observable, makeObservable } from 'mobx';

import {
  OpenAPIParameter,
  OpenAPIParameterLocation,
  OpenAPIParameterStyle,
  Referenced,
} from '../../types';
import { RedocNormalizedOptions } from '../RedocNormalizedOptions';

import { extractExtensions } from '../../utils/openapi';
import { OpenAPIParser } from '../OpenAPIParser';
import { SchemaModel } from './Schema';
import { ExampleModel } from './Example';
import { mapValues } from '../../utils/helpers';

const DEFAULT_SERIALIZATION: Record<
  OpenAPIParameterLocation,
  { explode: boolean; style: OpenAPIParameterStyle }
> = {
  path: {
    style: 'simple',
    explode: false,
  },
  query: {
    style: 'form',
    explode: true,
  },
  header: {
    style: 'simple',
    explode: false,
  },
  cookie: {
    style: 'form',
    explode: true,
  },
};

/**
 * Field or Parameter model ready to be used by components
 */
export class FieldModel {
  @observable
  expanded: boolean | undefined = false;

  schema: SchemaModel;
  name: string;
  required: boolean;
  description: string;
  example?: string;
  examples?: Record<string, ExampleModel>;
  deprecated: boolean;
  in?: OpenAPIParameterLocation;
  kind: string;
  extensions?: Record<string, any>;
  explode: boolean;
  style?: OpenAPIParameterStyle;
  const?: any;

  serializationMime?: string;

  constructor(
    parser: OpenAPIParser,
    infoOrRef: Referenced<OpenAPIParameter> & { name?: string; kind?: string },
    pointer: string,
    options: RedocNormalizedOptions,
  ) {
    makeObservable(this);

    const info = parser.deref<OpenAPIParameter>(infoOrRef);
    this.kind = infoOrRef.kind || 'field';
    this.name = infoOrRef.name || info.name;
    this.in = info.in;
    this.required = !!info.required;

    let fieldSchema = info.schema;
    let serializationMime = '';
    if (!fieldSchema && info.in && info.content) {
      serializationMime = Object.keys(info.content)[0];
      fieldSchema = info.content[serializationMime] && info.content[serializationMime].schema;
    }

    this.schema = new SchemaModel(parser, fieldSchema || {}, pointer, options);
    this.description =
      info.description === undefined ? this.schema.description || '' : info.description;
    this.example = info.example || this.schema.example;

    if (info.examples !== undefined) {
      this.examples = mapValues(
        info.examples,
        (example, name) => new ExampleModel(parser, example, name, info.encoding),
      );
    }

    if (serializationMime) {
      this.serializationMime = serializationMime;
    } else if (info.style) {
      this.style = info.style;
    } else if (this.in) {
      this.style = DEFAULT_SERIALIZATION[this.in]?.style ?? 'form'; // fallback to from in case "in" is invalid
    }

    if (info.explode === undefined && this.in) {
      this.explode = DEFAULT_SERIALIZATION[this.in]?.explode ?? true;
    } else {
      this.explode = !!info.explode;
    }

    this.deprecated = info.deprecated === undefined ? !!this.schema.deprecated : info.deprecated;
    parser.exitRef(infoOrRef);

    if (options.showExtensions) {
      this.extensions = extractExtensions(info, options.showExtensions);
    }

    this.const = this.schema?.const || info?.const || '';
  }

  @action
  toggle() {
    this.expanded = !this.expanded;
  }
}
