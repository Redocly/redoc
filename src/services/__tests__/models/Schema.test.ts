/* eslint-disable @typescript-eslint/no-var-requires */
import { parseYaml } from '@redocly/openapi-core';
import { outdent } from 'outdent';
import { MediaTypeModel } from '../../models';
import { SchemaModel } from '../../models/Schema';
import { OpenAPIParser } from '../../OpenAPIParser';
import { RedocNormalizedOptions } from '../../RedocNormalizedOptions';
import { enumDetailsPrinter, printSchema } from './helpers';

const opts = new RedocNormalizedOptions({});

describe('Models', () => {
  describe('Schema', () => {
    let parser;

    test('discriminator with one field', () => {
      const spec = require('../fixtures/discriminator.json');
      parser = new OpenAPIParser(spec, undefined, opts);
      const schema = new SchemaModel(parser, spec.components.schemas.Foo, '', opts);
      expect(schema.oneOf).toHaveLength(1);
      expect(schema.discriminatorProp).toEqual('type');
    });

    test('oneOf/allOf titles', () => {
      const spec = require('../fixtures/oneOfTitles.json');
      parser = new OpenAPIParser(spec, undefined, opts);
      const schema = new SchemaModel(parser, spec.components.schemas.Test, '', opts);
      expect(schema.fields).toHaveLength(3);
      const oneOfField = schema.fields![0];
      expect(oneOfField.schema.displayType).toBe('Foo (object) or Bar (object)');
      expect(oneOfField.schema.oneOf![0].title).toBe('Foo');
      expect(oneOfField.schema.oneOf![1].title).toBe('Bar');

      const anyOfField = schema.fields![1];
      expect(anyOfField.schema.displayType).toBe('Foo (object) or Bar (object)');
      expect(anyOfField.schema.oneOf![0].title).toBe('Foo');
      expect(anyOfField.schema.oneOf![1].title).toBe('Bar');
    });

    test('oneOf/allOf schema complex displayType', () => {
      const spec = require('../fixtures/oneOfTitles.json');
      parser = new OpenAPIParser(spec, undefined, opts);
      const schema = new SchemaModel(parser, spec.components.schemas.WithArray, '', opts);
      expect(schema.oneOf).toHaveLength(2);
      expect(schema.displayType).toBe('(Array of strings or numbers) or string');
    });

    test('schemaDefinition should resolve double ref', () => {
      const spec = require('../fixtures/3.1/schemaDefinition.json');
      parser = new OpenAPIParser(spec, undefined, opts);
      const schema = new SchemaModel(parser, spec.components.schemas.Parent, '', opts);
      expect(schema.fields).toHaveLength(1);
      expect(schema.pointer).toBe('#/components/schemas/Child');
    });

    test('schemaDefinition should resolve schema with conditional operators', () => {
      const spec = require('../fixtures/3.1/conditionalSchema.json');
      parser = new OpenAPIParser(spec, undefined, opts);
      const schema = new SchemaModel(parser, spec.components.schemas.Test, '', opts);
      expect(schema.oneOf).toHaveLength(2);

      expect(schema.oneOf![0].schema.title).toBe('=== 10');
      expect(schema.oneOf![1].schema.title).toBe('case 2');

      expect(schema.oneOf![0].schema).toMatchSnapshot();
      expect(schema.oneOf![1].schema).toMatchSnapshot();
    });

    test('schemaDefinition should resolve field with conditional operators', () => {
      const spec = require('../fixtures/3.1/conditionalField.json');
      parser = new OpenAPIParser(spec, undefined, opts);
      const schema = new SchemaModel(parser, spec.components.schemas.Test, '', opts);
      expect(schema.fields).toHaveLength(1);
      expect(schema.fields && schema.fields[0].schema.oneOf).toHaveLength(2);
      expect(schema.fields && schema.fields[0].schema.oneOf![0].schema.title).toBe('isString');
      expect(schema.fields && schema.fields[0].schema.oneOf![1].schema.title).toBe('notString');

      expect(schema.fields && schema.fields[0].schema.oneOf![0].schema).toMatchSnapshot();
      expect(schema.fields && schema.fields[0].schema.oneOf![1].schema).toMatchSnapshot();
    });

    test('schemaDefinition should resolve unevaluatedProperties in properties', () => {
      const spec = require('../fixtures/3.1/unevaluatedProperties.json');
      parser = new OpenAPIParser(spec, undefined, opts);
      const schema = new SchemaModel(parser, spec.components.schemas.Test, '', opts);
      expect(schema.fields).toHaveLength(2);
      expect(schema.fields![1].kind).toEqual('additionalProperties');
      expect(schema.fields![1].schema.type).toEqual('any');
    });

    test('schemaDefinition should resolve unevaluatedProperties in anyOf', () => {
      const spec = require('../fixtures/3.1/unevaluatedProperties.json');
      parser = new OpenAPIParser(spec, undefined, opts);
      const schema = new SchemaModel(parser, spec.components.schemas.Test2, '', opts);
      expect(schema.oneOf![0].fields).toHaveLength(2);
      expect(schema.oneOf![0].fields![1].kind).toEqual('additionalProperties');
      expect(schema.oneOf![1].fields).toHaveLength(2);
      expect(schema.oneOf![1].fields![1].kind).toEqual('additionalProperties');
    });

    test('schemaDefinition should resolve unevaluatedProperties type boolean', () => {
      const spec = require('../fixtures/3.1/unevaluatedProperties.json');
      parser = new OpenAPIParser(spec, undefined, opts);
      const schema = new SchemaModel(parser, spec.components.schemas.Test3, '', opts);
      expect(schema.fields).toHaveLength(2);
      expect(schema.fields![1].kind).toEqual('additionalProperties');
      expect(schema.fields![1].schema.type).toEqual('boolean');
    });

    test('schemaDefinition should resolve patternProperties', () => {
      const spec = require('../fixtures/3.1/patternProperties.json');
      parser = new OpenAPIParser(spec, undefined, opts);
      const schema = new SchemaModel(parser, spec.components.schemas.Patterns, '', opts);

      expect(schema.fields).toHaveLength(4);
      expect(schema.fields![0].kind).toEqual('field');
      expect(schema.fields![0].name).toEqual('nestedObjectProp');
      expect(schema.fields![0].schema.type).toEqual('object');
      expect(schema.fields![0].schema.fields![0].kind).toEqual('patternProperties');

      expect(schema.fields).toHaveLength(4);
      expect(schema.fields![1].kind).toEqual('field');
      expect(schema.fields![1].name).toEqual('nestedArrayProp');
      expect(schema.fields![1].schema.items!.fields![0].kind).toEqual('patternProperties');

      expect(schema.fields![2].kind).toEqual('patternProperties');
      expect(schema.fields![2].schema.type).toEqual('string');
      expect(schema.fields![3].kind).toEqual('patternProperties');
      expect(schema.fields![3].schema.type).toEqual('object');
    });

    describe('type array', () => {
      function testImmutablePart(schema: SchemaModel) {
        expect(schema.minItems).toEqual(1);
        expect(schema.maxItems).toEqual(10);
        expect(schema.fields![0].schema.type).toEqual('string');
        expect(schema.fields![1].schema.type).toEqual('number');
      }
      const eachArray = ['../fixtures/3.1/prefixItems.json', '../fixtures/arrayItems.json'];

      test.each(eachArray)(
        'schemaDefinition should resolve prefixItems without additional items',
        specFixture => {
          const spec = require(specFixture);
          const parser = new OpenAPIParser(spec, undefined, opts);
          const schema = new SchemaModel(parser, spec.components.schemas.Case1, '', opts);

          testImmutablePart(schema);

          expect(schema.fields).toHaveLength(3);
          expect(schema.fields![2].name).toEqual('[2]');
          expect(schema.fields![2].schema.pointer).toEqual('#/components/schemas/Cat');
          expect(schema.fields![2].schema.type).toEqual('object');
        },
      );

      test.each(eachArray)(
        'schemaDefinition should resolve prefixItems with additional items',
        specFixture => {
          const spec = require(specFixture);
          const parser = new OpenAPIParser(spec, undefined, opts);
          const schema = new SchemaModel(parser, spec.components.schemas.Case2, '', opts);

          testImmutablePart(schema);

          expect(schema.fields).toHaveLength(4);
          expect(schema.fields![3].name).toEqual('[3...]');
          expect(schema.fields![2].schema.type).toEqual('object');
          expect(schema.fields![2].schema.pointer).toEqual('#/components/schemas/Cat');
          expect(schema.fields![3].schema.type).toEqual('any');
        },
      );

      test.each(eachArray)(
        'schemaDefinition should resolve prefixItems with additional items with $ref',
        specFixture => {
          const spec = require(specFixture);
          const parser = new OpenAPIParser(spec, undefined, opts);
          const schema = new SchemaModel(parser, spec.components.schemas.Case3, '', opts);

          testImmutablePart(schema);

          expect(schema.fields).toHaveLength(4);
          expect(schema.fields![3].name).toEqual('[3...]');
          expect(schema.fields![2].schema.type).toEqual('object');
          expect(schema.fields![2].schema.pointer).toEqual('#/components/schemas/Cat');
          expect(schema.fields![3].schema.type).toEqual('object');
          expect(schema.fields![3].schema.pointer).toEqual('#/components/schemas/Dog');
        },
      );

      test.each(eachArray)(
        'schemaDefinition should resolve prefixItems with additional schema items',
        specFixture => {
          const spec = require(specFixture);
          const parser = new OpenAPIParser(spec, undefined, opts);
          const schema = new SchemaModel(parser, spec.components.schemas.Case4, '', opts);

          testImmutablePart(schema);

          expect(schema.fields).toHaveLength(4);
          expect(schema.fields![3].name).toEqual('[3...]');
          expect(schema.fields![2].schema.type).toEqual('object');
          expect(schema.fields![2].schema.pointer).toEqual('#/components/schemas/Cat');
          expect(schema.fields![3].schema.type).toEqual('object');
        },
      );

      test.each(eachArray)(
        'schemaDefinition should resolve prefixItems with additional array items',
        specFixture => {
          const spec = require(specFixture);
          const parser = new OpenAPIParser(spec, undefined, opts);
          const schema = new SchemaModel(parser, spec.components.schemas.Case5, '', opts);

          testImmutablePart(schema);

          expect(schema.fields).toHaveLength(4);
          expect(schema.fields![3].name).toEqual('[3...]');
          expect(schema.fields![2].schema.type).toEqual('object');
          expect(schema.fields![2].schema.pointer).toEqual('#/components/schemas/Cat');
          expect(schema.fields![3].schema.type).toEqual('array');
          expect(schema.fields![3].schema.fields).toHaveLength(1);
          expect(schema.fields![3].schema.fields![0].schema.type).toEqual('string');
          expect(schema.fields![3].schema.fields![0].schema.constraints).toEqual([
            '>= 0 characters',
          ]);
        },
      );

      test.each(eachArray)(
        'schemaDefinition should resolve prefixItems with additional array items',
        specFixture => {
          const spec = require(specFixture);
          const parser = new OpenAPIParser(spec, undefined, opts);
          const schema = new SchemaModel(parser, spec.components.schemas.Case6, '', opts);
          expect(schema.type).toBe('array');
          expect(schema.typePrefix).toBe('Array of ');
          expect(schema.items?.fields).toHaveLength(2);
          expect(schema.items?.pointer).toEqual('#/components/schemas/Tag');
          expect(schema.isPrimitive).toBe(false);
          expect(schema.items?.isPrimitive).toBe(false);
          expect(schema.minItems).toBe(1);
        },
      );

      test.each(eachArray)(
        'schemaDefinition should resolve items with boolean type',
        specFixture => {
          const spec = require(specFixture);
          const parser = new OpenAPIParser(spec, undefined, opts);
          const schema = new SchemaModel(parser, spec.components.schemas.Case7, '', opts);
          expect(schema.fields?.[0].schema?.type).toBe('array');
          expect(schema.fields?.[0].schema?.typePrefix).toBe('Array of ');
          expect(schema.fields?.[0].schema.items?.displayType).toBe('any');
          expect(schema?.fields).toHaveLength(1);
          expect(schema.fields?.[0].schema.pointer).toEqual('#/components/schemas/AnyArray');
          expect(schema.fields?.[0].schema.isPrimitive).toBe(true);
        },
      );
    });

    test('should get correct fields data if it includes allOf', () => {
      const spec = parseYaml(outdent`
      openapi: 3.0.0
      components:
        schemas:
          User:
            allOf:
              - type: object
                properties:
                  name:
                    type: string
                    description: correct description name
                    readOnly: false
                    writeOnly: false
                  allOf:
                    - '#/components/schemas/NameField'
          NameField:
            type: object
            description: name description
            readOnly: true
            writeOnly: false

      `) as any;

      parser = new OpenAPIParser(spec, undefined, opts);
      const schema = new SchemaModel(
        parser,
        spec.components.schemas.User,
        '#/components/schemas/User',
        opts,
      );
      const fieldSchema = schema.fields?.[0].schema;
      expect(fieldSchema?.readOnly).toBe(false);
      expect(fieldSchema?.writeOnly).toBe(false);
      expect(printSchema(schema)).toMatchInlineSnapshot(`
        "name: <string> (correct description name)
        allOf: <any>"
      `);
    });
    describe('enum values', () => {
      test('should get correct fields enum fields without duplication', () => {
        const spec = parseYaml(outdent`
          openapi: 3.0.0
          components:
            schemas:
              StringField: { type: string, title: StringField, enum: [A, B, C] }
              FieldA: { type: string, title: FieldA, enum: [A1, A2, A3] }
              FieldB: { type: string, title: FieldB, enum: [B1, B2, B3] }
              FieldC: { type: string, title: FieldC, enum: [C1, C2, C3] }
              ObjectWithAllOf:
                title: StringFilter
                type: object
                allOf:
                  - properties:
                      type: { type: string, enum: [STRING] }
                      field: { $ref: '#/components/schemas/StringField' }
                    required: [type, field, values]
                  - oneOf:
                      - properties:
                          field: { type: string, enum: [A] }
                          values: { type: array, items: { $ref: '#/components/schemas/FieldA' } }
                      - properties:
                          field: { type: string, enum: [B] }
                          values: { type: array, items: { $ref: '#/components/schemas/FieldB' } }
                      - properties:
                          field: { type: string, enum: [C] }
                          values: { type: array, items: { $ref: '#/components/schemas/FieldC' } }
              ObjectWithOneOf:
                title: StringFilter
                type: object
                properties:
                  type: { type: string, enum: [STRING] }
                  field: { $ref: '#/components/schemas/StringField' }
                required: [type, field, values]
                oneOf:
                  - properties:
                      field: { type: string, enum: [A] }
                      values: { type: array, items: { $ref: '#/components/schemas/FieldA' } }
                  - properties:
                      field: { type: string, enum: [B] }
                      values: { type: array, items: { $ref: '#/components/schemas/FieldB' } }
                  - properties:
                      field: { type: string, enum: [C] }
                      values: { type: array, items: { $ref: '#/components/schemas/FieldC' } }
        `) as any;

        parser = new OpenAPIParser(spec, undefined, opts);
        const schemaWithOneOf = new SchemaModel(
          parser,
          spec.components.schemas.ObjectWithOneOf,
          '#/components/schemas/ObjectWithOneOf',
          opts,
        );
        expect(printSchema(schemaWithOneOf, enumDetailsPrinter)).toMatchInlineSnapshot(`
          "oneOf
            StringFilter ->
              field*: <string>enum: [A,B,C]
              values*: [<string>enum: [A1,A2,A3]]
              type*: <string>enum: [STRING]
            StringFilter ->
              field*: <string>enum: [A,B,C]
              values*: [<string>enum: [B1,B2,B3]]
              type*: <string>enum: [STRING]
            StringFilter ->
              field*: <string>enum: [A,B,C]
              values*: [<string>enum: [C1,C2,C3]]
              type*: <string>enum: [STRING]"
        `);

        const schemaWithAllOf = new SchemaModel(
          parser,
          spec.components.schemas.ObjectWithAllOf,
          '#/components/schemas/ObjectWithAllOf',
          opts,
        );
        expect(printSchema(schemaWithAllOf, enumDetailsPrinter)).toMatchInlineSnapshot(`
          "oneOf
            object ->
              type*: <string>enum: [STRING]
              field*: <string>enum: [A,B,C]
              values*: [<string>enum: [A1,A2,A3]]
            object ->
              type*: <string>enum: [STRING]
              field*: <string>enum: [B,A,C]
              values*: [<string>enum: [B1,B2,B3]]
            object ->
              type*: <string>enum: [STRING]
              field*: <string>enum: [C,A,B]
              values*: [<string>enum: [C1,C2,C3]]"
        `);
      });

      test('should get correct fields enum limits', () => {
        const spec = parseYaml(outdent`
          openapi: 3.0.0
          components:
            schemas:
              StringField: { type: string, title: StringField, enum: [A, B, C] }
              FieldA: { type: string, title: FieldA, enum: [A1, A2, A3] }
              FieldB: { type: string, title: FieldB, enum: [B1, B2, B3] }
              FieldC: { type: string, title: FieldC, enum: [C1, C2, C3] }
              ObjectWithAllOf:
                title: StringFilter
                type: object
                allOf:
                  - properties:
                      type: { type: string, enum: [STRING] }
                    required: [type, field, values]
                  - oneOf:
                      - properties:
                          field: { type: string, enum: [A] }
                          values: { type: array, items: { $ref: '#/components/schemas/FieldA' } }
                      - properties:
                          field: { type: string, enum: [B] }
                          values: { type: array, items: { $ref: '#/components/schemas/FieldB' } }
                      - properties:
                          field: { type: string, enum: [C] }
                          values: { type: array, items: { $ref: '#/components/schemas/FieldC' } }
              ObjectWithOneOf:
                title: StringFilter
                type: object
                properties:
                  type: { type: string, enum: [STRING] }
                required: [type, field, values]
                oneOf:
                  - properties:
                      field: { type: string, enum: [A] }
                      values: { type: array, items: { $ref: '#/components/schemas/FieldA' } }
                  - properties:
                      field: { type: string, enum: [B] }
                      values: { type: array, items: { $ref: '#/components/schemas/FieldB' } }
                  - properties:
                      field: { type: string, enum: [C] }
                      values: { type: array, items: { $ref: '#/components/schemas/FieldC' } }
        `) as any;

        parser = new OpenAPIParser(spec, undefined, opts);
        const schemaWithOneOf = new SchemaModel(
          parser,
          spec.components.schemas.ObjectWithOneOf,
          '#/components/schemas/ObjectWithOneOf',
          opts,
        );
        expect(printSchema(schemaWithOneOf, enumDetailsPrinter)).toMatchInlineSnapshot(`
          "oneOf
            StringFilter ->
              field*: <string>enum: [A]
              values*: [<string>enum: [A1,A2,A3]]
              type*: <string>enum: [STRING]
            StringFilter ->
              field*: <string>enum: [B]
              values*: [<string>enum: [B1,B2,B3]]
              type*: <string>enum: [STRING]
            StringFilter ->
              field*: <string>enum: [C]
              values*: [<string>enum: [C1,C2,C3]]
              type*: <string>enum: [STRING]"
        `);

        const schemaWithAllOf = new SchemaModel(
          parser,
          spec.components.schemas.ObjectWithAllOf,
          '#/components/schemas/ObjectWithAllOf',
          opts,
        );
        expect(printSchema(schemaWithAllOf, enumDetailsPrinter)).toMatchInlineSnapshot(`
          "oneOf
            object ->
              type*: <string>enum: [STRING]
              field*: <string>enum: [A]
              values*: [<string>enum: [A1,A2,A3]]
            object ->
              type*: <string>enum: [STRING]
              field*: <string>enum: [B]
              values*: [<string>enum: [B1,B2,B3]]
            object ->
              type*: <string>enum: [STRING]
              field*: <string>enum: [C]
              values*: [<string>enum: [C1,C2,C3]]"
        `);
      });
    });

    test('should get correct sibling inside schema type for openapi 3.1', () => {
      const spec = parseYaml(outdent`
        openapi: 3.1.0
        paths:
          /test:
            get:
              operationId: test
              responses:
                '200':
                  content:
                    application/json:
                      schema:
                        type: object
                        properties:
                          testAttr:
                            description: Overridden description
                            type: string
                            $ref: '#/components/schemas/Test'
        components:
          schemas:
            Test:
              type: object
              description: Refed description
      `) as any;

      parser = new OpenAPIParser(spec, undefined, opts);
      const name = 'application/json';
      const mediaType = new MediaTypeModel(
        parser,
        name,
        true,
        spec.paths['/test'].get.responses['200'].content[name],
        opts,
      );

      expect(printSchema(mediaType?.schema as any)).toMatchInlineSnapshot(
        `"testAttr: <string> (Overridden description)"`,
      );
    });

    test('should not override schema in openapi 3.0', () => {
      const spec = parseYaml(outdent`
        openapi: 3.0.0
        paths:
          /test:
            get:
              operationId: test
              responses:
                '200':
                  content:
                    application/json:
                      schema:
                        type: object
                        properties:
                          testAttr:
                            type: string
                            description: Overridden description
                            $ref: '#/components/schemas/Test'
        components:
          schemas:
            Test:
              type: object
              description: Refed description
      `) as any;

      parser = new OpenAPIParser(spec, undefined, opts);
      const name = 'application/json';
      const mediaType = new MediaTypeModel(
        parser,
        name,
        true,
        spec.paths['/test'].get.responses['200'].content[name],
        opts,
      );

      expect(printSchema(mediaType?.schema as any)).toMatchInlineSnapshot(
        `"testAttr: <object> (Refed description)"`,
      );
    });
  });
});
