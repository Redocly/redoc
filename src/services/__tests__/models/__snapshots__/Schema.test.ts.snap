// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Models Schema schemaDefinition should resolve field with conditional operators 1`] = `
Object {
  "allOf": undefined,
  "default": undefined,
  "description": undefined,
  "items": Object {
    "allOf": undefined,
    "description": undefined,
    "format": "url",
    "readOnly": undefined,
    "title": undefined,
    "type": "string",
    "writeOnly": undefined,
    "x-circular-ref": undefined,
    "x-parentRefs": Array [],
  },
  "maxItems": 20,
  "minItems": 1,
  "readOnly": undefined,
  "title": "isString",
  "type": "string",
  "writeOnly": undefined,
  "x-circular-ref": undefined,
  "x-displayName": "isString",
  "x-parentRefs": Array [],
}
`;

exports[`Models Schema schemaDefinition should resolve field with conditional operators 2`] = `
Object {
  "allOf": undefined,
  "default": undefined,
  "description": undefined,
  "items": Object {
    "allOf": undefined,
    "description": undefined,
    "format": "url",
    "readOnly": undefined,
    "title": undefined,
    "type": "string",
    "writeOnly": undefined,
    "x-circular-ref": undefined,
    "x-parentRefs": Array [],
  },
  "maxItems": 10,
  "minItems": 1,
  "pattern": "\\\\d+",
  "readOnly": undefined,
  "title": "notString",
  "type": Array [
    "string",
    "integer",
    "null",
  ],
  "writeOnly": undefined,
  "x-circular-ref": undefined,
  "x-displayName": "notString",
  "x-parentRefs": Array [],
}
`;

exports[`Models Schema schemaDefinition should resolve schema with conditional operators 1`] = `
Object {
  "allOf": undefined,
  "description": undefined,
  "maxItems": 2,
  "properties": Object {
    "test": Object {
      "allOf": undefined,
      "description": "The list of URL to a cute photos featuring pet",
      "enum": Array [
        10,
      ],
      "items": Object {
        "allOf": undefined,
        "description": undefined,
        "format": "url",
        "readOnly": undefined,
        "title": undefined,
        "type": "string",
        "writeOnly": undefined,
        "x-circular-ref": undefined,
        "x-parentRefs": Array [],
      },
      "maxItems": 20,
      "minItems": 1,
      "readOnly": undefined,
      "title": undefined,
      "type": Array [
        "string",
        "integer",
        "null",
      ],
      "writeOnly": undefined,
      "x-circular-ref": undefined,
      "x-parentRefs": Array [],
      "x-refsStack": Array [
        "/oneOf/0",
      ],
    },
  },
  "readOnly": undefined,
  "title": "=== 10",
  "type": "object",
  "writeOnly": undefined,
  "x-circular-ref": undefined,
  "x-parentRefs": Array [],
}
`;

exports[`Models Schema schemaDefinition should resolve schema with conditional operators 2`] = `
Object {
  "allOf": undefined,
  "description": undefined,
  "maxItems": 20,
  "properties": Object {
    "test": Object {
      "description": "The list of URL to a cute photos featuring pet",
      "items": Object {
        "format": "url",
        "type": "string",
      },
      "maxItems": 20,
      "minItems": 1,
      "type": Array [
        "string",
        "integer",
        "null",
      ],
      "x-refsStack": Array [
        "/oneOf/1",
      ],
    },
  },
  "readOnly": undefined,
  "title": "case 2",
  "type": "object",
  "writeOnly": undefined,
  "x-circular-ref": undefined,
  "x-parentRefs": Array [],
}
`;
